### ISSUE TITLE
Problem(Argentina): Can't provide past data to TN

### ISSUE THREAD

**Issue by outerlook** [trufnetwork/adapters#24]
Spec says:

> The oldest available date is 14 Aug 2024

but in reality we can only fetch data of the last 7 days, as the documents gets updated 

![image](https://github.com/user-attachments/assets/bc34b2fe-ea5a-4382-b1cf-a9401472a1c0)
here each weekday name is mentioned (Jueves, etc)

![image](https://github.com/user-attachments/assets/2fb2058d-734c-481f-bc6e-6bab75c3ad01)
here it's said:
> Se actualizó el recurso [Viernes](https://datos.produccion.gob.ar/dataset/6f47ec76-d1ce-4e34-a7e1-621fe9b1d0b5/archivo/91bc072a-4726-44a1-85ec-4a8467aad27e) en el dataset [Precios Claros - Base SEPA](https://datos.produccion.gob.ar/dataset/sepa-precios). Hace 6 días.

in english:
> The resource [Friday](https://datos.produccion.gob.ar/dataset/6f47ec76-d1ce-4e34-a7e1-621fe9b1d0b5/archivo/91bc072a-4726-44a1-85ec-4a8467aad27e) in the dataset [Precios Claros - Base SEPA](https://datos.produccion.gob.ar/dataset/sepa-precios) was updated. 6 days ago.

so it gets overwritten, and I can't access it back once it's updated.

**Comment by outerlook** [comment]
hey @markholdex, do you know better about how this was asked? is it expected just to solve from now on, or was it expected to have all data?

do we have a connection from the gov to clarify this, or should I try the contact points on the website?

**Comment by markholdex** [comment]
@outerlook previous data will be provided once we have the current implementation ready. We can even handle it in a separate Goal. The reason is that we need a mechanism in ready. Once ready, we can begin supplying old ZIP files.

**Comment by outerlook** [comment]
Alright! 

One thing isn't clear. Current mechanism includes scraping the website. 

It's easy to change that, but what would be the final mechanism? Have a list of files ready to be fed and discovered? Or partial from website and partial from this list?

**Comment by outerlook** [comment]
Alright! 

One thing isn't clear. Current mechanism includes scraping the website. 

It's easy to change that, but what would be the final mechanism? Have a list of files ready to be fed and discovered? Or partial from website and partial from this list?

Or all of these can be delayed to next goal?

**Comment by markholdex** [comment]
@outerlook the final mechanism is the following: 

- We need to create an endpoint where we can upload (ideally the main ZIP file directly) or the csv with the prices. From there, once the file is uploaded, the adapter will begin processing our data. 
- The endpoint should be obviously secured and restricted to entities that we allow to use it.

**Comment by rsoury** [comment]
> We need to create an endpoint where we can upload (ideally the main ZIP file directly) or the csv with the prices. From there, once the file is uploaded, the adapter will begin processing our data.

I agree. I've recommended an internal S3 bucket owned by Truflation, that is private, and restricted to specific AWS org Ids and configured for egress to be paid by the downloader.
Helium Network takes this approach for their data pipeline.

**Comment by outerlook** [comment]
@markholdex @rsoury 
I'll try to see if we can simplify this, but I guess a valid workflow could be using S3 + pre-signed URLs for upload 

Pre-signed URLs are important to restrict access to who can upload to a S3 owned by Truflation. Downside: additional step to get pre-signed URLs, they can't live more than 1 week.

steps:
1. Provider sends a GET to a URL protected by some token we share (protection doesn't need to be too fancy here, for stakes of this vector I suppose), and the file name they intend to upload (needed for pre-signed URL)
2. lambda processes the request, responds with the URL
3. they use it to upload on their side, programmatically

We could make our pipeline push-based instead of the current pull-based approach with S3 events, which could be more efficient. However, if immediate processing is not a requirement, I'll keep the current pipeline, using the bucket instead of the website to poll for data. This will help us ship faster. Also increasing the schedule to check more frequently (each hour maybe)

All looks good? Any suggestion/concern is important

**Comment by rsoury** [comment]
Can't we just add Org Ids to access permissions on S3?
If we can avoid Lambda function that'd be awesome.

Otherwise, @markholdex - this should meet the requirements you've stated - unless you have a simpler approach.

**Comment by markholdex** [comment]
@outerlook @rsoury I think if we can just have a shared S3 bucket, that would be more than enough. We assume that Arg. Gov is a trusted entity and they won't be uploading random stuff there except the ZIP we need. So maybe we try to go with this approach?

**Comment by outerlook** [comment]
Alright! Ryan's idea works fine, I just checked and S3 management cross-org sharing is all good. I'll create a directory on a private S3 bucket and grant list/write/read access to their account once we know which one.

**Comment by markholdex** [comment]
@outerlook 

1. is it mandatory to have an instruction around the naming convention? 
2. Also, can we already begin to upload past data?

**Comment by outerlook** [comment]
@markholdex 
sorry, I missed this notification.

> 1. is it mandatory to have an instruction around the naming convention?

This greatly simplifies things and edge cases, it's better to enforce, unless there's a hard blockers that prevents from adopting this convention. For now, the convention is this: 
```
s3://argentina-sepa/source_data/sepa_2025-01-03.zip
    ^^^             ^^^^^       ^^^^^^
    bucket          prefix      sepa_{date}.zip
```

> 2. Also, can we already begin to upload past data?

Yes, here:

https://us-east-2.console.aws.amazon.com/s3/buckets/argentina-sepa?prefix=source_data/&region=us-east-2&bucketType=general

note that there are already some data I've been populating to test it. But it's already free to add official data.

**Comment by markholdex** [comment]
@outerlook and our process will automatically begin to review and ingest data?

https://us-east-2.console.aws.amazon.com/s3/buckets/argentina-sepa?prefix=source_data/&region=us-east-2&bucketType=general

Will you be able to record a demo for me?

**Comment by outerlook** [comment]
after deployment, yes. The caveat:

I didn't implement a filling behavior to reconciliate data (to simplify development). It checks for new data after last inserted for each stream. For now, if we ever have some very old data needed to be inserted, the easiest is drop all and redeploy

of course I can deploy some more sophisticated filling behavior later if it's a requirement, but it's just better not to run hourly as the main flow will.

**Comment by outerlook** [comment]
Hey @markholdex 
here's a demo: https://www.youtube.com/watch?v=rzt0DQ_Nwgs

it both explains and show it working (locally). wasn't sure what is the target intention, but use the description to jump to [02:22](https://www.youtube.com/watch?v=rzt0DQ_Nwgs&t=142s) Upload a new date if it's just to show how to upload.

if it's something else let me know. I did this quick & dirty just to be fast

**Comment by Ivan-TRUF** [comment]
@markholdex @outerlook the goal is to fetch only new data every day and store it in database with historical values staying the same, what is needed to accomplish that? In addition to that, we need to fetch ALL the historical data they have available, and that is I believe since Aug 2024

**Comment by markholdex** [comment]
> the goal is to fetch only new data every day and store it in database with historical values staying the same, what is needed to accomplish that? 

@Ivan-TRUF not sure what is confusing you but this is already happening as part of the process. In the Demo video that Raffael presented, it demonstrates how an automated process is fetching data presented in a zip file and pushes it to TN with the ability to read that data through CLI/SDK. 

> In addition to that, we need to fetch ALL the historical data they have available, and that is I believe since Aug 2024. I thought that this is clear from specs provided at the beginning of this project?


Since we don't have all historical data yet, and the Argentina governance is yet to provide it, we built a process where the zip file can just be dropped in our data bucket and it will be processed. It's why we have this GH issue opened until we don't insert all historical data.

**Comment by Ivan-TRUF** [comment]
@markholdex and what is the earliest date that we have the data from?

**Comment by markholdex** [comment]
@Ivan-TRUF I have a zip file on my local machine from Nov 28th. But that doesn't mean we have the files for every day after that. If we want chronologically consistent data on the production, we can ingest only the data from the current week as this is the earliest available to devs.

**Comment by Ivan-TRUF** [comment]
@markholdex not sure what you have but we should definitely have data for every day since inception, and onwards, we just have to look at the latest data, no need to check the whole week. Does this make sense?
Cc @outerlook

**Comment by markholdex** [comment]
@Ivan-TRUF do you have all the zip files from the inception day?

**Comment by Ivan-TRUF** [comment]
They will provide them, at least I hope they have it but we should have a system that stores files every day. Do we have that? @markholdex

**Comment by Ivan-TRUF** [comment]
@markholdex great! Is it operational, aka running continuously? If yes, what is the earliest file that we have? But also all the files after that

**Comment by Ivan-TRUF** [comment]
@markholdex that sounds good! Once it's live it might be good to go over that during call to see if it all looks good!
